fix(cli): disable API interaction when running in offline mode

byCedric opened the pull request:
# Why

Fixes ENG-6550
Fixes #18918

# How

- Disabled dependency validation when running in offline mode
- Disabled Expo Go version validation when running in offline mode
  > **Note** it still checks if Expo Go is installed. When it's not installed, it throws instead.
- Disabled fetching user when running in offline mode
  > **Note** this would fall back to an anonymous user and (auto)disables analytics and (possible) manifest signing. 
- Updated tests to validate dependency validation is skipped in offline mode

Optionally, we could add a global catch for errors listed below, to mention the `--offline` option.
 - `error instanceof CommandError && error.code === 'API'`
 - `error.code === 'FetchError'`

# Test Plan

To test this, update your **/etc/hosts** and add `127.0.0.1 api.expo.dev`. Then run the CLI:
- `npx expo start` → this fails with a fetch error
- `npx expo start --offline` → this should succeed without issues
- `npx expo start --offline --android|ios` 
  → when Expo Go is NOT installed, it should throw
  → when Expo Go is installed, it should warn and continue to the bundle

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

byCedric committed:
fix(cli): disable dependency validation when running in offline mode

byCedric commented on packages/@expo/cli/src/start/doctor/dependencies/validateDependenciesVersions.ts:
```diff
@@ -32,7 +34,12 @@ export async function validateDependenciesVersionsAsync(
   packagesToCheck?: string[]
-): Promise<boolean> {
+): Promise<boolean | null> {
+  if (APISettings.isOffline) {
```
I added the offline check here, but since `expo start` is the only one supporting the `--offline` flag, we could also not-run this function if running in offline mode.

No strong opinions on either one, but with this, we know that this won't pop up again if used in other commands.

byCedric committed:
docs(cli): add changelog entry

EvanBacon commented on packages/@expo/cli/src/start/doctor/dependencies/validateDependenciesVersions.ts:
```diff
@@ -32,7 +34,12 @@ export async function validateDependenciesVersionsAsync(
-): Promise<boolean> {
+): Promise<boolean | null> {
+  if (APISettings.isOffline) {
+    debug('Skipping dependency validation in offline mode');
```
```suggestion
    Log.warn('Skipping dependency validation in offline mode');
```

byCedric committed:
refactor(cli): show warning about skipped dependency validation

Co-authored-by: Evan Bacon <bacon@expo.io>

byCedric committed:
fix(cli): Add support for offline mode in Expo Go installer

byCedric committed:
fix(cli): Skip fetching user when running in offline mode

This will default to an anonymous user, and should skip the signed manifest part. Resulting in just being able to load the bundle.

byCedric committed:
docs(cli): update changelog entry

byCedric committed:
test(cli): drop unused CommandError import from Expo Go installer

byCedric commented on packages/@expo/cli/src/start/platforms/ExpoGoInstaller.ts:
```diff
@@ -74,6 +76,16 @@ export class ExpoGoInstaller<IDevice> {
+      Log.warn(`Skipping Expo Go version validation in offline mode`);
+      return false;
+    } else if (APISettings.isOffline && shouldInstall) {
+      throw new CommandError(
+        'NO_EXPO_GO',
+        `Expo Go is not installed on device "${deviceManager.name}", while running in offline mode. Manually install Expo Go or run without --offline flag.`
+      );
```
I think this is the best context we can give them. Pointing to an `expo.fyi` seems pointless in offline mode's case 🤷

Stevemoretz commented:
Is this published yet? I updated but didn't get it yet.

brentvatne commented:
@Stevemoretz - published in expo@46.0.16
